src/protego.py

```diff
@@ -0,0 +1,248 @@
+import re
+from datetime import time
+from collections import namedtuple
+from six.moves.urllib.parse import urlparse, urlunparse, ParseResult, quote, unquote
+
+Record = namedtuple('Record', ['field', 'value'])
+RequestRate = namedtuple(
+    'RequestRate', ['requests', 'seconds', 'start_time', 'end_time'])
+
+DISALLOW_DIRECTIVE = ['disallow', 'dissallow', 'dissalow', 'disalow', 'diasllow', 'disallaw']
+ALLOW_DIRECTIVE = ['allow']
+USER_AGENT_DIRECTIVE = ['user-agent', 'useragent', 'user agent']
+SITEMAP_DIRECTIVE = ['sitemap', 'sitemaps', 'site-map']
+CRAWL_DELAY_DIRECTIVE = ['crawl-delay', 'crawl delay']
+REQUEST_RATE_DIRECTIVE = ['request-rate', 'request rate']
+
+
+def is_valid_directive(d):
+    return d in DISALLOW_DIRECTIVE or d in ALLOW_DIRECTIVE \
+        or d in USER_AGENT_DIRECTIVE or d in SITEMAP_DIRECTIVE \
+        or d in CRAWL_DELAY_DIRECTIVE or d in REQUEST_RATE_DIRECTIVE
+
+
+class RecordGroup:
+
+    def __init__(self):
+        self.user_agents = []
+        self.rules = []
+        self.crawl_delay = None
+        self.sorted_rules = None
+        self.req_rate = None
+
+    def add_user_agent(self, user_agent):
+        user_agent = user_agent.strip().lower()
+        if user_agent != '*' and '*' in user_agent:
+            user_agent = user_agent.replace('*', '')
```

Comment:
I would leave it for a separate change after the initial commit, but I think that, when a `*` is found in the middle of a user agent string, the intention of the `robots.txt` author has likely been for it to match any character or a literal asterisk, instead of being ignored.

Reply:
No current parser (including google's parser) support wildcards in user agent values. From the data we collected using top 1000 websites, I could see that few robots.txt have user agents with `*` at the end, which really doesn't make any difference. Since google's parser just ignores the asterisk, I thought Protego should do the same.